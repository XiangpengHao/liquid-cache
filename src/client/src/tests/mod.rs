use std::{path::Path, sync::Arc};

use datafusion::{
    physical_plan::{ExecutionPlan, display::DisplayableExecutionPlan},
    prelude::{SessionConfig, SessionContext},
};

use crate::LiquidCacheBuilder;

async fn setup_tpch_ctx() -> Arc<SessionContext> {
    let mut session_config = SessionConfig::from_env().unwrap();
    // To get deterministic results
    session_config.options_mut().execution.target_partitions = 8;
    let ctx = LiquidCacheBuilder::new("http://localhost:50051")
        .build(session_config)
        .unwrap();
    let tables = [
        "customer", "lineitem", "nation", "orders", "part", "partsupp", "region", "supplier",
    ];

    for table in tables {
        let table_url = format!("../../benchmark/tpch/data/sf0.001/{}.parquet", table);
        ctx.register_parquet(table, table_url, Default::default())
            .await
            .unwrap();
    }
    Arc::new(ctx)
}

/// One query file can contain multiple queries, separated by `;`
fn get_query_by_id(query_dir: impl AsRef<Path>, query_id: u32) -> Vec<String> {
    let query_dir = query_dir.as_ref();
    let mut path = query_dir.to_owned();
    path.push(format!("q{query_id}.sql"));
    let content = std::fs::read_to_string(&path).unwrap();
    content
        .split(';')
        .map(|s| s.trim())
        .filter(|s| !s.is_empty())
        .map(|s| s.to_string())
        .collect()
}

async fn get_query_plan_by_id(query_id: u32) -> Arc<dyn ExecutionPlan> {
    let query_dir = Path::new("../../benchmark/tpch/queries");
    let ctx = setup_tpch_ctx().await;

    let query = if query_id != 15 {
        &get_query_by_id(query_dir, query_id)[0]
    } else {
        &get_query_by_id(query_dir, query_id)[1]
    };
    let result = ctx.sql(query).await.unwrap();
    let (state, plan) = result.into_parts();
    state.create_physical_plan(&plan).await.unwrap()
}

#[tokio::test]
async fn test_tpch_q1() {
    let physical_plan = get_query_plan_by_id(1).await;
    let displayable = DisplayableExecutionPlan::new(physical_plan.as_ref());
    insta::assert_snapshot!(displayable.tree_render().to_string());
}

#[tokio::test]
async fn test_tpch_q2() {
    let physical_plan = get_query_plan_by_id(2).await;
    let displayable = DisplayableExecutionPlan::new(physical_plan.as_ref());
    insta::assert_snapshot!(displayable.tree_render().to_string());
}

#[tokio::test]
async fn test_tpch_q3() {
    let physical_plan = get_query_plan_by_id(3).await;
    let displayable = DisplayableExecutionPlan::new(physical_plan.as_ref());
    insta::assert_snapshot!(displayable.tree_render().to_string());
}

#[tokio::test]
async fn test_tpch_q4() {
    let physical_plan = get_query_plan_by_id(4).await;
    let displayable = DisplayableExecutionPlan::new(physical_plan.as_ref());
    insta::assert_snapshot!(displayable.tree_render().to_string());
}

#[tokio::test]
async fn test_tpch_q5() {
    let physical_plan = get_query_plan_by_id(5).await;
    let displayable = DisplayableExecutionPlan::new(physical_plan.as_ref());
    insta::assert_snapshot!(displayable.tree_render().to_string());
}

#[tokio::test]
async fn test_tpch_q6() {
    let physical_plan = get_query_plan_by_id(6).await;
    let displayable = DisplayableExecutionPlan::new(physical_plan.as_ref());
    insta::assert_snapshot!(displayable.tree_render().to_string());
}

#[tokio::test]
async fn test_tpch_q7() {
    let physical_plan = get_query_plan_by_id(7).await;
    let displayable = DisplayableExecutionPlan::new(physical_plan.as_ref());
    insta::assert_snapshot!(displayable.tree_render().to_string());
}

#[tokio::test]
async fn test_tpch_q8() {
    let physical_plan = get_query_plan_by_id(8).await;
    let displayable = DisplayableExecutionPlan::new(physical_plan.as_ref());
    insta::assert_snapshot!(displayable.tree_render().to_string());
}

#[tokio::test]
async fn test_tpch_q9() {
    let physical_plan = get_query_plan_by_id(9).await;
    let displayable = DisplayableExecutionPlan::new(physical_plan.as_ref());
    insta::assert_snapshot!(displayable.tree_render().to_string());
}

#[tokio::test]
async fn test_tpch_q10() {
    let physical_plan = get_query_plan_by_id(10).await;
    let displayable = DisplayableExecutionPlan::new(physical_plan.as_ref());
    insta::assert_snapshot!(displayable.tree_render().to_string());
}

#[tokio::test]
async fn test_tpch_q11() {
    let physical_plan = get_query_plan_by_id(11).await;
    let displayable = DisplayableExecutionPlan::new(physical_plan.as_ref());
    insta::assert_snapshot!(displayable.tree_render().to_string());
}

#[tokio::test]
async fn test_tpch_q12() {
    let physical_plan = get_query_plan_by_id(12).await;
    let displayable = DisplayableExecutionPlan::new(physical_plan.as_ref());
    insta::assert_snapshot!(displayable.tree_render().to_string());
}

#[tokio::test]
async fn test_tpch_q13() {
    let physical_plan = get_query_plan_by_id(13).await;
    let displayable = DisplayableExecutionPlan::new(physical_plan.as_ref());
    insta::assert_snapshot!(displayable.tree_render().to_string());
}

#[tokio::test]
async fn test_tpch_q14() {
    let physical_plan = get_query_plan_by_id(14).await;
    let displayable = DisplayableExecutionPlan::new(physical_plan.as_ref());
    insta::assert_snapshot!(displayable.tree_render().to_string());
}

#[tokio::test]
/// Q15 needs to actually run the first query to get the correct plan
#[ignore]
async fn test_tpch_q15() {
    let physical_plan = get_query_plan_by_id(15).await;
    let displayable = DisplayableExecutionPlan::new(physical_plan.as_ref());
    insta::assert_snapshot!(displayable.tree_render().to_string());
}

#[tokio::test]
async fn test_tpch_q16() {
    let physical_plan = get_query_plan_by_id(16).await;
    let displayable = DisplayableExecutionPlan::new(physical_plan.as_ref());
    insta::assert_snapshot!(displayable.tree_render().to_string());
}

#[tokio::test]
async fn test_tpch_q17() {
    let physical_plan = get_query_plan_by_id(17).await;
    let displayable = DisplayableExecutionPlan::new(physical_plan.as_ref());
    insta::assert_snapshot!(displayable.tree_render().to_string());
}

#[tokio::test]
async fn test_tpch_q18() {
    let physical_plan = get_query_plan_by_id(18).await;
    let displayable = DisplayableExecutionPlan::new(physical_plan.as_ref());
    insta::assert_snapshot!(displayable.tree_render().to_string());
}

#[tokio::test]
async fn test_tpch_q19() {
    let physical_plan = get_query_plan_by_id(19).await;
    let displayable = DisplayableExecutionPlan::new(physical_plan.as_ref());
    insta::assert_snapshot!(displayable.tree_render().to_string());
}

#[tokio::test]
async fn test_tpch_q20() {
    let physical_plan = get_query_plan_by_id(20).await;
    let displayable = DisplayableExecutionPlan::new(physical_plan.as_ref());
    insta::assert_snapshot!(displayable.tree_render().to_string());
}

#[tokio::test]
async fn test_tpch_q21() {
    let physical_plan = get_query_plan_by_id(21).await;
    let displayable = DisplayableExecutionPlan::new(physical_plan.as_ref());
    insta::assert_snapshot!(displayable.tree_render().to_string());
}

#[tokio::test]
async fn test_tpch_q22() {
    let physical_plan = get_query_plan_by_id(22).await;
    let displayable = DisplayableExecutionPlan::new(physical_plan.as_ref());
    insta::assert_snapshot!(displayable.tree_render().to_string());
}
