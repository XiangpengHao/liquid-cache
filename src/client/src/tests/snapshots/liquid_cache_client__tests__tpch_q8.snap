---
source: src/client/src/tests/mod.rs
expression: displayable.tree_render().to_string()
---
┌───────────────────────────┐
│  SortPreservingMergeExec  │
│    --------------------   │
│   o_year ASC NULLS LAST   │
└─────────────┬─────────────┘
┌─────────────┴─────────────┐
│          SortExec         │
│    --------------------   │
│  o_year@0 ASC NULLS LAST  │
└─────────────┬─────────────┘
┌─────────────┴─────────────┐
│       ProjectionExec      │
│    --------------------   │
│         mkt_share:        │
│ sum(CASE WHEN all_nations │
│  .nation = Utf8("BRAZIL") │
│   THEN all_nations.volume │
│    ELSE Int64(0) END) /   │
│       sum(all_nations     │
│          .volume)         │
│                           │
│       o_year: o_year      │
└─────────────┬─────────────┘
┌─────────────┴─────────────┐
│       AggregateExec       │
│    --------------------   │
│           aggr:           │
│ , sum(all_nations.volume) │
│                           │
│      group_by: o_year     │
│                           │
│           mode:           │
│      FinalPartitioned     │
└─────────────┬─────────────┘
┌─────────────┴─────────────┐
│    CoalesceBatchesExec    │
│    --------------------   │
│     target_batch_size:    │
│           16384           │
└─────────────┬─────────────┘
┌─────────────┴─────────────┐
│      RepartitionExec      │
│    --------------------   │
│ partition_count(in->out): │
│           8 -> 8          │
│                           │
│    partitioning_scheme:   │
│    Hash([o_year@0], 8)    │
└─────────────┬─────────────┘
┌─────────────┴─────────────┐
│       AggregateExec       │
│    --------------------   │
│           aggr:           │
│ , sum(all_nations.volume) │
│                           │
│      group_by: o_year     │
│       mode: Partial       │
└─────────────┬─────────────┘
┌─────────────┴─────────────┐
│       ProjectionExec      │
│    --------------------   │
│       nation: n_name      │
│                           │
│          o_year:          │
│      date_part(YEAR,      │
│        o_orderdate)       │
│                           │
│          volume:          │
│ l_extendedprice * (Some(1)│
│    ,20,0 - l_discount)    │
└─────────────┬─────────────┘
┌─────────────┴─────────────┐
│    CoalesceBatchesExec    │
│    --------------------   │
│     target_batch_size:    │
│           16384           │
└─────────────┬─────────────┘
┌─────────────┴─────────────┐
│        HashJoinExec       │
│    --------------------   │
│            on:            ├──────────────┐
│(r_regionkey = n_regionkey)│              │
└─────────────┬─────────────┘              │
┌─────────────┴─────────────┐┌─────────────┴─────────────┐
│   LiquidCacheClientExec   ││       ProjectionExec      │
│    --------------------   ││    --------------------   │
│          server:          ││        l_discount:        │
│  http://localhost:50051,  ││         l_discount        │
│        mode=liquid,       ││                           │
│      object_stores=[]     ││      l_extendedprice:     │
│                           ││      l_extendedprice      │
│                           ││                           │
│                           ││       n_name: n_name      │
│                           ││                           │
│                           ││        n_regionkey:       │
│                           ││        n_regionkey        │
│                           ││                           │
│                           ││        o_orderdate:       │
│                           ││        o_orderdate        │
└─────────────┬─────────────┘└─────────────┬─────────────┘
┌─────────────┴─────────────┐┌─────────────┴─────────────┐
│       DataSourceExec      ││    CoalesceBatchesExec    │
│    --------------------   ││    --------------------   │
│          files: 1         ││     target_batch_size:    │
│      format: parquet      ││           16384           │
│                           ││                           │
│         predicate:        ││                           │
│      r_name = AMERICA     ││                           │
└───────────────────────────┘└─────────────┬─────────────┘
                             ┌─────────────┴─────────────┐
                             │        HashJoinExec       │
                             │    --------------------   │
                             │            on:            ├──────────────┐
                             │(n_nationkey = s_nationkey)│              │
                             └─────────────┬─────────────┘              │
                             ┌─────────────┴─────────────┐┌─────────────┴─────────────┐
                             │   LiquidCacheClientExec   ││       ProjectionExec      │
                             │    --------------------   ││    --------------------   │
                             │          server:          ││        l_discount:        │
                             │  http://localhost:50051,  ││         l_discount        │
                             │        mode=liquid,       ││                           │
                             │      object_stores=[]     ││      l_extendedprice:     │
                             │                           ││      l_extendedprice      │
                             │                           ││                           │
                             │                           ││        n_regionkey:       │
                             │                           ││        n_regionkey        │
                             │                           ││                           │
                             │                           ││        o_orderdate:       │
                             │                           ││        o_orderdate        │
                             │                           ││                           │
                             │                           ││        s_nationkey:       │
                             │                           ││        s_nationkey        │
                             └─────────────┬─────────────┘└─────────────┬─────────────┘
                             ┌─────────────┴─────────────┐┌─────────────┴─────────────┐
                             │       DataSourceExec      ││    CoalesceBatchesExec    │
                             │    --------------------   ││    --------------------   │
                             │          files: 1         ││     target_batch_size:    │
                             │      format: parquet      ││           16384           │
                             └───────────────────────────┘└─────────────┬─────────────┘
                                                          ┌─────────────┴─────────────┐
                                                          │        HashJoinExec       │
                                                          │    --------------------   │
                                                          │            on:            ├──────────────┐
                                                          │(n_nationkey = c_nationkey)│              │
                                                          └─────────────┬─────────────┘              │
                                                          ┌─────────────┴─────────────┐┌─────────────┴─────────────┐
                                                          │   LiquidCacheClientExec   ││       ProjectionExec      │
                                                          │    --------------------   ││    --------------------   │
                                                          │          server:          ││        c_nationkey:       │
                                                          │  http://localhost:50051,  ││        c_nationkey        │
                                                          │        mode=liquid,       ││                           │
                                                          │      object_stores=[]     ││        l_discount:        │
                                                          │                           ││         l_discount        │
                                                          │                           ││                           │
                                                          │                           ││      l_extendedprice:     │
                                                          │                           ││      l_extendedprice      │
                                                          │                           ││                           │
                                                          │                           ││        o_orderdate:       │
                                                          │                           ││        o_orderdate        │
                                                          │                           ││                           │
                                                          │                           ││        s_nationkey:       │
                                                          │                           ││        s_nationkey        │
                                                          └─────────────┬─────────────┘└─────────────┬─────────────┘
                                                          ┌─────────────┴─────────────┐┌─────────────┴─────────────┐
                                                          │       DataSourceExec      ││    CoalesceBatchesExec    │
                                                          │    --------------------   ││    --------------------   │
                                                          │          files: 1         ││     target_batch_size:    │
                                                          │      format: parquet      ││           16384           │
                                                          └───────────────────────────┘└─────────────┬─────────────┘
                                                                                       ┌─────────────┴─────────────┐
                                                                                       │        HashJoinExec       │
                                                                                       │    --------------------   │
                                                                                       │            on:            ├──────────────┐
                                                                                       │  (c_custkey = o_custkey)  │              │
                                                                                       └─────────────┬─────────────┘              │
                                                                                       ┌─────────────┴─────────────┐┌─────────────┴─────────────┐
                                                                                       │   LiquidCacheClientExec   ││       ProjectionExec      │
                                                                                       │    --------------------   ││    --------------------   │
                                                                                       │          server:          ││        l_discount:        │
                                                                                       │  http://localhost:50051,  ││         l_discount        │
                                                                                       │        mode=liquid,       ││                           │
                                                                                       │      object_stores=[]     ││      l_extendedprice:     │
                                                                                       │                           ││      l_extendedprice      │
                                                                                       │                           ││                           │
                                                                                       │                           ││    o_custkey: o_custkey   │
                                                                                       │                           ││                           │
                                                                                       │                           ││        o_orderdate:       │
                                                                                       │                           ││        o_orderdate        │
                                                                                       │                           ││                           │
                                                                                       │                           ││        s_nationkey:       │
                                                                                       │                           ││        s_nationkey        │
                                                                                       └─────────────┬─────────────┘└─────────────┬─────────────┘
                                                                                       ┌─────────────┴─────────────┐┌─────────────┴─────────────┐
                                                                                       │       DataSourceExec      ││    CoalesceBatchesExec    │
                                                                                       │    --------------------   ││    --------------------   │
                                                                                       │          files: 1         ││     target_batch_size:    │
                                                                                       │      format: parquet      ││           16384           │
                                                                                       └───────────────────────────┘└─────────────┬─────────────┘
                                                                                                                    ┌─────────────┴─────────────┐
                                                                                                                    │        HashJoinExec       │
                                                                                                                    │    --------------------   │
                                                                                                                    │            on:            ├──────────────┐
                                                                                                                    │ (o_orderkey = l_orderkey) │              │
                                                                                                                    └─────────────┬─────────────┘              │
                                                                                                                    ┌─────────────┴─────────────┐┌─────────────┴─────────────┐
                                                                                                                    │   LiquidCacheClientExec   ││       ProjectionExec      │
                                                                                                                    │    --------------------   ││    --------------------   │
                                                                                                                    │          server:          ││        l_discount:        │
                                                                                                                    │  http://localhost:50051,  ││         l_discount        │
                                                                                                                    │        mode=liquid,       ││                           │
                                                                                                                    │      object_stores=[]     ││      l_extendedprice:     │
                                                                                                                    │                           ││      l_extendedprice      │
                                                                                                                    │                           ││                           │
                                                                                                                    │                           ││        l_orderkey:        │
                                                                                                                    │                           ││         l_orderkey        │
                                                                                                                    │                           ││                           │
                                                                                                                    │                           ││        s_nationkey:       │
                                                                                                                    │                           ││        s_nationkey        │
                                                                                                                    └─────────────┬─────────────┘└─────────────┬─────────────┘
                                                                                                                    ┌─────────────┴─────────────┐┌─────────────┴─────────────┐
                                                                                                                    │       DataSourceExec      ││    CoalesceBatchesExec    │
                                                                                                                    │    --------------------   ││    --------------------   │
                                                                                                                    │          files: 1         ││     target_batch_size:    │
                                                                                                                    │      format: parquet      ││           16384           │
                                                                                                                    │                           ││                           │
                                                                                                                    │         predicate:        ││                           │
                                                                                                                    │ o_orderdate >= 1995-01-01 ││                           │
                                                                                                                    │   AND o_orderdate <= 1996 ││                           │
                                                                                                                    │           -12-31          ││                           │
                                                                                                                    └───────────────────────────┘└─────────────┬─────────────┘
                                                                                                                                                 ┌─────────────┴─────────────┐
                                                                                                                                                 │        HashJoinExec       │
                                                                                                                                                 │    --------------------   │
                                                                                                                                                 │            on:            ├──────────────┐
                                                                                                                                                 │  (s_suppkey = l_suppkey)  │              │
                                                                                                                                                 └─────────────┬─────────────┘              │
                                                                                                                                                 ┌─────────────┴─────────────┐┌─────────────┴─────────────┐
                                                                                                                                                 │   LiquidCacheClientExec   ││      RepartitionExec      │
                                                                                                                                                 │    --------------------   ││    --------------------   │
                                                                                                                                                 │          server:          ││ partition_count(in->out): │
                                                                                                                                                 │  http://localhost:50051,  ││           1 -> 8          │
                                                                                                                                                 │        mode=liquid,       ││                           │
                                                                                                                                                 │      object_stores=[]     ││    partitioning_scheme:   │
                                                                                                                                                 │                           ││     RoundRobinBatch(8)    │
                                                                                                                                                 └─────────────┬─────────────┘└─────────────┬─────────────┘
                                                                                                                                                 ┌─────────────┴─────────────┐┌─────────────┴─────────────┐
                                                                                                                                                 │       DataSourceExec      ││    CoalesceBatchesExec    │
                                                                                                                                                 │    --------------------   ││    --------------------   │
                                                                                                                                                 │          files: 1         ││     target_batch_size:    │
                                                                                                                                                 │      format: parquet      ││           16384           │
                                                                                                                                                 └───────────────────────────┘└─────────────┬─────────────┘
                                                                                                                                                                              ┌─────────────┴─────────────┐
                                                                                                                                                                              │        HashJoinExec       │
                                                                                                                                                                              │    --------------------   │
                                                                                                                                                                              │            on:            ├──────────────┐
                                                                                                                                                                              │  (p_partkey = l_partkey)  │              │
                                                                                                                                                                              └─────────────┬─────────────┘              │
                                                                                                                                                                              ┌─────────────┴─────────────┐┌─────────────┴─────────────┐
                                                                                                                                                                              │   LiquidCacheClientExec   ││   LiquidCacheClientExec   │
                                                                                                                                                                              │    --------------------   ││    --------------------   │
                                                                                                                                                                              │          server:          ││          server:          │
                                                                                                                                                                              │  http://localhost:50051,  ││  http://localhost:50051,  │
                                                                                                                                                                              │        mode=liquid,       ││        mode=liquid,       │
                                                                                                                                                                              │      object_stores=[]     ││      object_stores=[]     │
                                                                                                                                                                              └─────────────┬─────────────┘└─────────────┬─────────────┘
                                                                                                                                                                              ┌─────────────┴─────────────┐┌─────────────┴─────────────┐
                                                                                                                                                                              │       DataSourceExec      ││       DataSourceExec      │
                                                                                                                                                                              │    --------------------   ││    --------------------   │
                                                                                                                                                                              │          files: 1         ││          files: 1         │
                                                                                                                                                                              │      format: parquet      ││      format: parquet      │
                                                                                                                                                                              │                           ││                           │
                                                                                                                                                                              │         predicate:        ││                           │
                                                                                                                                                                              │ p_type = ECONOMY ANODIZED ││                           │
                                                                                                                                                                              │            STEEL          ││                           │
                                                                                                                                                                              └───────────────────────────┘└───────────────────────────┘
